test/fw/ptl/utils/pbs_dshutils.py

```diff
@@ -2070,55 +2070,80 @@ def create_temp_dir(self, hostname=None, suffix='', prefix='PtlPbs',
-        :type dir: str or None
-        :param uid: Optional username or uid of temp directory owner
-        :param gid: Optional group name or gid of temp directory
+        :param dirname: the directory will be created in this directory
+        :type dirname: str or None
+        :param asuser: Optional username of temp directory owner
+        :type asuser: str
+        :param asgroup: Optional group name of temp directory
                     group owner
+        :type asgroup: str
         :param mode: Optional mode bits to assign to the temporary
                      directory
+        :type mode: octal integer
         :param level: logging level, defaults to INFOCLI2
         """
+        current_user_info = self.get_id_info(self.get_current_user())
+        uid = current_user_info['uid']
+        gid = current_user_info['gid']
+        if asuser is not None:
+            uid = self.get_id_info(str(asuser))['uid']
+        if asgroup is not None:
+            gid = grp.getgrnam(str(asgroup))[2]
         # create a temp dir as current user
         tmpdir = tempfile.mkdtemp(suffix, prefix)
         if dirname is not None:
             dirname = str(dirname)
+            self.chmod(path=tmpdir, mode=0o755)
             self.run_copy(hostname, src=tmpdir, dest=dirname, runas=asuser,
-                          recursive=True,
-                          preserve_permission=False, level=level)
+                          recursive=True, gid=gid, uid=uid,
+                          level=level)
+            if mode is not None:
+                self.chmod(hostname, path=dirname, mode=mode, runas=asuser)
+            else:
+                self.chmod(hostname, path=dirname, mode=0o755, runas=asuser)
+
             tmpdir = dirname + tmpdir[4:]
 
         # if temp dir to be created on remote host
         if not self.is_localhost(hostname):
             if asuser is not None:
-                # by default mkstemp creates dir with 0600 permission
+                # by default mkstemp creates dir with 0o700 permission
                 # to create dir as different user first change the dir
-                # permission to 0644 so that other user has read permission
+                # permission to 0o755 so that other user has read permission
                 self.chmod(path=tmpdir, mode=0o755)
                 # copy temp dir created on local host to remote host
                 # as different user
                 self.run_copy(hostname, src=tmpdir, dest=tmpdir, runas=asuser,
-                              recursive=True,
-                              preserve_permission=False, level=level)
+                              recursive=True, uid=uid, gid=gid,
+                              level=level)
+                if mode is not None:
+                    self.chmod(hostname, path=tmpdir, mode=mode, runas=asuser)
+                else:
+                    self.chmod(hostname, path=tmpdir, mode=0o755, runas=asuser)
             else:
```

Comment:
looks like we are repeating the code which we have inside if. Why don't we place the check just before the last chmod?
if chmod is a NOOP when asuser is None, this check may not be even required.

All three conditions in the function do a somewhat similar sequence of operations. Can it be combined to form a generic sequence without explicit conditions?

Reply:
Right, While that can be done but it will affect readability a lot. I'll have to set src and destination early up in the function to then call common lines of the code and then again based on the three conditions cleanup unnecessary directories.
I am anyways making a change to reduce the number of chmod calls, that should help make it smaller.